projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/control/LLVMDispatchBasicBlockNode.java

```diff
@@ -93,11 +98,36 @@ public Object executeGeneric(VirtualFrame frame) {
         int backEdgeCounter = 0;
         outer: while (basicBlockIndex != LLVMBasicBlockNode.RETURN_FROM_FUNCTION) {
             CompilerAsserts.partialEvaluationConstant(basicBlockIndex);
-            LLVMBasicBlockNode bb = bodyNodes[basicBlockIndex];
+
+            if(bodyNodes[basicBlockIndex] instanceof LLVMLoopNode) {
+                LLVMLoopNode loop = (LLVMLoopNode) bodyNodes[basicBlockIndex];
+                loop.execute(frame);
+
+                Integer[] successors = loop.getSuccessors();
+                for(int i = 0; i < successors.length; i++) {
+                    try {
+                        if(frame.getInt(loopSuccessorSlot) == successors[i]) {
+                            basicBlockIndex = successors[i];
+                            continue outer;
+                        }
+                    } catch (FrameSlotTypeException e) {
+                        CompilerDirectives.transferToInterpreter();
+                        throw new RuntimeException("Error while reading from loop successor frame slot - type mismatch!");
+                    }
+                }
+
+                CompilerDirectives.transferToInterpreter();
+                throw new IllegalStateException("Must not reach here!");
```

Comment:
Why not? You could make this error message more descriptive by e.g. stating that you did not find a loop successor.